### PR TITLE
Refactor packagemanager lines

### PR DESCRIPTION
**Pull request by KevinRansom**
In order to add support for #i, first I have done some refactoring around packagemanager lines.

The updates include a discriminator indicating #i or #r , although perhaps a discriminated union would work better.

And adds a few utility helpers to PackageManagerLine to manage the maps.

### CODE REVIEW

File: src/fsharp/CompileOps.fsi [hunk 1]
```diff
@@ -254,6 +254,26 @@ type VersionFlag =
...
+    | Unprocessed
+    | Processed
+
+type PackageManagerLine =
+    { LineType: LType
+      LineStatus: LStatus
+      Line: string
+      Range: range }
+
+    with
```
Comment (cartermp) [inline src/fsharp/CompileOps.fsi:271]: Same as impl file, this should not be required I think

File: src/fsharp/CompileOps.fs [hunk 2]
```diff
@@ -2086,6 +2086,62 @@ type CcuLoadFailureAction =
...
+    | Processed
+
+type PackageManagerLine =
+    { LineType: LType
+      LineStatus: LStatus
+      Line: string
+      Range: range }
+
+    with
+        static member AddLineWithKey (packageKey: string) (lt:LType) (line: string) (m: range) (packageManagerLines: Map<string, PackageManagerLine list>): Map<string, PackageManagerLine list>  =
```
[the remark this answers is not in the document]
Comment (cartermp) [reply src/fsharp/CompileOps.fs:2104]: I would think not given the interface file

File: src/fsharp/fsi/fsi.fs [hunk 3]
```diff
@@ -1444,54 +1444,55 @@ type internal FsiDynamicCompiler
...
         if not needsPackageResolution then istate else
         needsPackageResolution <- false
 
         tcConfigB.packageManagerLines |> Seq.fold(fun istate kv ->
-            let inline snd3 (_, b, _) = b
             let packageManagerKey, packageManagerLines = kv.Key, kv.Value
             match packageManagerLines with
             | [] -> istate
-            | (_, _, m)::_ ->
+            | {LineType=_; LineStatus=_; Line=_; Range=m} :: _ ->
```
Comment (cartermp) [inline src/fsharp/fsi/fsi.fs:1460]:
```suggestion
            | {  LineType=_; LineStatus=_; Line=_; Range=m } :: _ ->
```

File: src/fsharp/fsi/fsi.fs [hunk 4]
```diff
@@ -1444,54 +1444,55 @@ type internal FsiDynamicCompiler
...
+                let reportError =
+                    let report errorType err msg =
+                        let error = err, msg
+                        match errorType with
+                        | ErrorReportType.Warning -> warning(Error(error, m))
+                        | ErrorReportType.Error -> errorR(Error(error, m))
+                    ResolvingErrorReport (report)
```
Comment (cartermp) [inline src/fsharp/fsi/fsi.fs:1467]: Ugh, did I add this from the merge somehow? :(

File: src/fsharp/CompileOps.fs [hunk 5]
```diff
@@ -2086,6 +2086,62 @@ type CcuLoadFailureAction =
...
+                packageManagerLines 
+                |> Map.map(fun key lines ->
+                    if key = packageKey then
+                        lines |> List.map(fun line -> {line with LineStatus = LStatus.Processed;})
+                    else
+                        lines)
+            map
+
+        static member StripDependencyManagerKey (packageKey: string) (line: string): string =
+            line.Substring(packageKey.Length + 1).Trim()
```
[the remark this answers is not in the document]
Comment (KevinRansom) [reply src/fsharp/CompileOps.fs:2143]: Package key can't have leading or trailing spaces, so trim just takes them away.
Somehow I prefer substring.

**Pull request merged**